File: go/init_contract_test.go
package haiai

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

type initBootstrapRegisterContract struct {
	Method            string `json:"method"`
	Path              string `json:"path"`
	AuthRequired      bool   `json:"auth_required"`
	PublicKeyEncoding string `json:"public_key_encoding"`
}

type registrationOutcomeCase struct {
	Name           string          `json:"name"`
	HTTPStatus     int             `json:"http_status"`
	Response       json.RawMessage `json:"response"`
	ExpectedStatus *string         `json:"expected_status"`
	ExpectedEmail  *string         `json:"expected_email"`
}

type initContractFixture struct {
	RegistrationOutcomes struct {
		Cases []registrationOutcomeCase `json:"cases"`
	} `json:"registration_outcomes"`
	BootstrapRegister        initBootstrapRegisterContract `json:"bootstrap_register"`
	ExistingIdentityRegister struct {
		Response json.RawMessage `json:"response"`
		Cases    []struct {
			Name    string            `json:"name"`
			Request map[string]string `json:"request"`
		} `json:"cases"`
	} `json:"existing_identity_register"`
	PrivateKeyCandidateOrder   []string `json:"private_key_candidate_order"`
	ConfigDiscoveryOrder       []string `json:"config_discovery_order"`
	PrivateKeyPasswordSources  []string `json:"private_key_password_sources"`
	PrivateKeyPasswordStrategy string   `json:"private_key_password_strategy"`
}

type registrationCaptureFFI struct {
	FFIClient // Any unexpected operation (including identity creation) fails.
	payloads  []string
	response  json.RawMessage
	err       error
}

func (f *registrationCaptureFFI) Register(optionsJSON string) (json.RawMessage, error) {
	f.payloads = append(f.payloads, optionsJSON)
	return f.response, f.err
}

func TestInitContractExistingIdentityRegister(t *testing.T) {
	contract := loadInitContractFixture(t).ExistingIdentityRegister
	if len(contract.Cases) == 0 {
		t.Fatal("missing existing-identity registration cases")
	}
	for _, tc := range contract.Cases {
		t.Run(tc.Name, func(t *testing.T) {
			ffi := &registrationCaptureFFI{response: contract.Response}
			client := &Client{ffi: ffi}
			result, err := client.Register(context.Background(), RegisterOptions{
				AgentJSON:       tc.Request["agent_json"],
				PublicKey:       tc.Request["public_key_pem"],
				OwnerEmail:      tc.Request["owner_email"],
				RegistrationKey: tc.Request["registration_key"],
			})
			if err != nil {
				t.Fatal(err)
			}
			if len(ffi.payloads) != 1 {
				t.Fatal("expected one existing-identity FFI registration call")
			}
			var payload map[string]string
			if err := json.Unmarshal([]byte(ffi.payloads[0]), &payload); err != nil {
				t.Fatal(err)
			}
			if !reflect.DeepEqual(payload, tc.Request) {
				t.Fatal("FFI registration payload does not match shared contract")
			}
			var expected RegistrationResult
			if err := json.Unmarshal(contract.Response, &expected); err != nil {
				t.Fatal(err)
			}
			if result.AgentID != expected.AgentID {
				t.Fatal("registration response was not preserved")
			}
		})
	}
}

func loadInitContractFixture(t *testing.T) initContractFixture {
	t.Helper()

	data, err := os.ReadFile("../fixtures/init_contract.json")
	if err != nil {
		t.Fatalf("read init contract fixture: %v", err)
	}

	var fixture initContractFixture
	if err := json.Unmarshal(data, &fixture); err != nil {
		t.Fatalf("decode init contract fixture: %v", err)
	}
	return fixture
}

func TestInitContractKeyCandidateOrder(t *testing.T) {
	fixture := loadInitContractFixture(t)
	expectedDiscovery := []string{"explicit_path", "JACS_CONFIG_PATH", "./jacs.config.json"}
	if strings.Join(fixture.ConfigDiscoveryOrder, "|") != strings.Join(expectedDiscovery, "|") {
		t.Fatalf("unexpected config discovery order: got %v, want %v", fixture.ConfigDiscoveryOrder, expectedDiscovery)
	}

	expectedPasswordSources := []string{"JACS_PRIVATE_KEY_PASSWORD", "JACS_PASSWORD_FILE"}
	if strings.Join(fixture.PrivateKeyPasswordSources, "|") != strings.Join(expectedPasswordSources, "|") {
		t.Fatalf("unexpected password source list: got %v, want %v", fixture.PrivateKeyPasswordSources, expectedPasswordSources)
	}
	if fixture.PrivateKeyPasswordStrategy != "single_source_required" {
		t.Fatalf("unexpected password strategy: got %q", fixture.PrivateKeyPasswordStrategy)
	}

	tmpDir := t.TempDir()
	keyDir := filepath.Join(tmpDir, "keys")
	if err := os.MkdirAll(keyDir, 0o755); err != nil {
		t.Fatalf("mkdir key dir: %v", err)
	}

	cfg := &Config{
		JacsAgentName: "agent-alpha",
		JacsKeyDir:    "./keys",
	}
	configPath := filepath.Join(tmpDir, "jacs.config.json")

	candidateNames := make([]string, 0, len(fixture.PrivateKeyCandidateOrder))
	for _, raw := range fixture.PrivateKeyCandidateOrder {
		candidateNames = append(candidateNames, strings.ReplaceAll(raw, "{agentName}", cfg.JacsAgentName))
	}
	candidatePaths := []string{
		filepath.Join(keyDir, candidateNames[0]),
		filepath.Join(keyDir, candidateNames[1]),
		filepath.Join(keyDir, candidateNames[2]),
	}

	for i, p := range candidatePaths {
		if err := os.WriteFile(p, []byte("key-"+candidateNames[i]), 0o600); err != nil {
			t.Fatalf("write candidate key: %v", err)
		}
	}

	got := ResolveKeyPath(cfg, configPath)
	if got != candidatePaths[0] {
		t.Fatalf("expected first candidate %q, got %q", candidatePaths[0], got)
	}

	if err := os.Remove(candidatePaths[0]); err != nil {
		t.Fatalf("remove first candidate: %v", err)
	}
	got = ResolveKeyPath(cfg, configPath)
	if got != candidatePaths[1] {
		t.Fatalf("expected second candidate %q, got %q", candidatePaths[1], got)
	}

	if err := os.Remove(candidatePaths[1]); err != nil {
		t.Fatalf("remove second candidate: %v", err)
	}
	got = ResolveKeyPath(cfg, configPath)
	if got != candidatePaths[2] {
		t.Fatalf("expected third candidate %q, got %q", candidatePaths[2], got)
	}
}

func TestInitContractBootstrapRegister(t *testing.T) {
	fixture := loadInitContractFixture(t)

	// With FFI, the Go SDK delegates bootstrap registration to Rust.
	// The mock FFI client's RegisterNewAgent posts to /api/v1/agents/register
	// on the httptest server (no auth header), matching the contract.
	var gotAuth string
	var gotMethod string
	var gotPath string

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotMethod = r.Method
		gotPath = r.URL.Path
		gotAuth = r.Header.Get("Authorization")

		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{
			"agent_id":"agent-123",
			"jacs_id":"jacs-123",
			"success":true,
			"dns_verified":false,
			"name":"agent-alpha",
			"public_key_path":"/tmp/pub.pem",
			"private_key_path":"/tmp/priv.pem",
			"config_path":"/tmp/jacs.config.json",
			"registrations":[]
		}`))
	}))
	defer srv.Close()

	cl, _ := newTestClient(t, srv.URL)
	_, err := cl.RegisterNewAgent(
		context.Background(),
		"agent-alpha",
		&RegisterNewAgentOptions{
			OwnerEmail:  "owner@hai.ai",
			Domain:      "agent.example",
			Description: "Go init contract",
			Password:    "test-password",
			Quiet:       true,
		},
	)
	if err != nil {
		t.Fatalf("RegisterNewAgent: %v", err)
	}

	if gotMethod != fixture.BootstrapRegister.Method {
		t.Fatalf("unexpected method: got %s, want %s", gotMethod, fixture.BootstrapRegister.Method)
	}
	if gotPath != fixture.BootstrapRegister.Path {
		t.Fatalf("unexpected path: got %s, want %s", gotPath, fixture.BootstrapRegister.Path)
	}
	if fixture.BootstrapRegister.AuthRequired && gotAuth == "" {
		t.Fatal("expected auth header but none was sent")
	}
	if !fixture.BootstrapRegister.AuthRequired && gotAuth != "" {
		t.Fatalf("expected no Authorization header, got %q", gotAuth)
	}
}

func (f *registrationCaptureFFI) RegisterNewAgent(optionsJSON string) (json.RawMessage, error) {
	f.payloads = append(f.payloads, optionsJSON)
	return f.response, f.err
}

func TestInitContractRegistrationOutcomes(t *testing.T) {
	for _, tc := range loadInitContractFixture(t).RegistrationOutcomes.Cases {
		for _, bootstrap := range []bool{false, true} {
			t.Run(fmt.Sprintf("%s/bootstrap=%t", tc.Name, bootstrap), func(t *testing.T) {
				ffi := &registrationCaptureFFI{response: tc.Response}
				if tc.HTTPStatus >= 400 {
					ffi.err = fmt.Errorf("API rejection: %s", tc.Response)
				}
				client := &Client{ffi: ffi}
				var status, email *string
				var agentID string
				var err error
				if bootstrap {
					var result *RegisterResult
					result, err = client.RegisterNewAgent(context.Background(), "requested-agent", &RegisterNewAgentOptions{
						OwnerEmail: "owner@example.test", Password: "synthetic-password", Quiet: true,
					})
					if result != nil {
						status, email, agentID = result.RegistrationStatus, result.Email, result.AgentID
					}
				} else {
					var result *RegistrationResult
					result, err = client.Register(context.Background(), RegisterOptions{AgentJSON: `{"jacsId":"existing"}`})
					if result != nil {
						status, email, agentID = result.RegistrationStatus, result.Email, result.AgentID
					}
				}
				if len(ffi.payloads) != 1 {
					t.Fatal("expected exactly one FFI call")
				}
				if tc.HTTPStatus >= 400 {
					if err == nil || !strings.Contains(err.Error(), "Registration key was not accepted") {
						t.Fatalf("expected registration rejection, got %v", err)
					}
					return
				}
				if err != nil {
					t.Fatal(err)
				}
				var response struct {
					AgentID string `json:"agent_id"`
				}
				if err := json.Unmarshal(tc.Response, &response); err != nil {
					t.Fatal(err)
				}
				if agentID != response.AgentID || !reflect.DeepEqual(status, tc.ExpectedStatus) || !reflect.DeepEqual(email, tc.ExpectedEmail) {
					t.Fatalf("registration outcome not preserved: status=%v email=%v agent=%s", status, email, agentID)
				}
			})
		}
	}
}

File: node/tests/init-contract.test.ts
import { afterEach, describe, expect, it, vi } from 'vitest';
import { mkdtemp, mkdir, rm, writeFile, unlink } from 'node:fs/promises';
import { readFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { tmpdir } from 'node:os';
import { HaiClient, registerNewAgent } from '../src/client.js';
import { loadConfig, loadPrivateKey } from '../src/config.js';
import { generateTestKeypair as generateKeypair } from './setup.js';
import { createMockFFI } from './ffi-mock.js';
import { FFIClientAdapter } from '../src/ffi-client.js';

interface BootstrapRegisterContract {
  method: string;
  path: string;
  auth_required: boolean;
  public_key_encoding: string;
}

interface InitContractFixture {
  bootstrap_register: BootstrapRegisterContract;
  registration_outcomes: {
    requested_name: string;
    cases: Array<{
      name: string;
      http_status: number;
      response: Record<string, unknown>;
      expected_status: string | null;
      expected_email: string | null;
    }>;
  };
  existing_identity_register: {
    response: { agent_id: string; jacs_id: string; registered_at: string };
    cases: Array<{
      name: string;
      request: { agent_json: string; owner_email: string; registration_key?: string; public_key_pem?: string };
    }>;
  };
  private_key_candidate_order: string[];
  config_discovery_order: string[];
  private_key_password_sources: string[];
  private_key_password_strategy: string;
}

function loadInitContractFixture(): InitContractFixture {
  const here = dirname(fileURLToPath(import.meta.url));
  const fixturePath = resolve(here, '../../fixtures/init_contract.json');
  return JSON.parse(readFileSync(fixturePath, 'utf-8')) as InitContractFixture;
}

describe('shared init contract (node)', () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  const registrationContract = loadInitContractFixture().existing_identity_register;
  it.each(registrationContract.cases)('existing identity registration: $name', async ({ request }) => {
    const keypair = generateKeypair();
    const client = await HaiClient.fromCredentials('fixture-existing-agent', keypair.privateKeyPem);
    const registerNative = vi.fn(async (_optionsJson: string) =>
      JSON.stringify(registrationContract.response),
    );
    // Exercise the real adapter's JSON serializer; only the native call is replaced.
    const adapter = Object.assign(Object.create(FFIClientAdapter.prototype), {
      native: { register: registerNative },
    }) as FFIClientAdapter;
    client._setFFIAdapter(adapter);

    const result = await client.register({
      agentJson: request.agent_json,
      ownerEmail: request.owner_email,
      ...(request.public_key_pem === undefined ? {} : { publicKeyPem: request.public_key_pem }),
      ...(request.registration_key === undefined ? {} : { registrationKey: request.registration_key }),
    });

    expect(registerNative).toHaveBeenCalledOnce();
    expect(JSON.parse(registerNative.mock.calls[0][0])).toEqual(request);
    expect(result.agentId).toBe(registrationContract.response.agent_id);
  });

  for (const entrypoint of ['ordinary', 'bootstrap', 'standalone'] as const) {
    it.each(loadInitContractFixture().registration_outcomes.cases)(
      `${entrypoint} preserves registration outcome: $name`, async (testCase) => {
        const nativeCall = vi.fn(async (_optionsJson: string) => {
          if (testCase.http_status >= 400) throw new Error(String(testCase.response.message));
          return JSON.stringify(testCase.response);
        });
        const adapter = Object.assign(Object.create(FFIClientAdapter.prototype), {
          native: { register: nativeCall, registerNewAgent: nativeCall },
        }) as FFIClientAdapter;
        const output = vi.spyOn(console, 'log').mockImplementation(() => {});
        vi.spyOn(FFIClientAdapter, 'create').mockResolvedValue(adapter);
        const opts = { ownerEmail: 'owner@example.test', password: 'synthetic-password' };
        const invoke = async () => {
          if (entrypoint === 'standalone') return registerNewAgent('requested-agent', opts);
          const client = await HaiClient.fromCredentials('existing', generateKeypair().privateKeyPem);
          client._setFFIAdapter(adapter);
          return entrypoint === 'ordinary'
            ? client.register({ agentJson: '{"jacsId":"existing"}' })
            : client.registerNewAgent('requested-agent', opts);
        };
        if (testCase.http_status >= 400) {
          await expect(invoke()).rejects.toThrow('Registration key was not accepted');
        } else {
          const result = await invoke();
          expect(result.agentId).toBe(testCase.response.agent_id);
          expect(result.registrationStatus).toBe(testCase.expected_status ?? undefined);
          expect(result.email).toBe(testCase.expected_email ?? undefined);
          if (entrypoint !== 'ordinary') {
            const text = output.mock.calls.map(args => args.join(' ')).join('\n');
            expect(text).toContain(`Registration status: ${testCase.expected_status ?? 'unknown'}`);
            expect(text).not.toContain('requested-agent@hai.ai');
            expect(text.toLowerCase()).not.toContain('verification email has been sent');
            if (testCase.expected_email) expect(text).toContain(`Assigned email: ${testCase.expected_email}`);
            else expect(text).not.toContain('Assigned email:');
          }
        }
        expect(nativeCall).toHaveBeenCalledOnce();
      },
    );
  }

  it('private key candidate order matches shared fixture', async () => {
    const fixture = loadInitContractFixture();
    expect(fixture.config_discovery_order).toEqual([
      'explicit_path',
      'JACS_CONFIG_PATH',
      './jacs.config.json',
    ]);
    expect(fixture.private_key_password_sources).toEqual([
      'JACS_PRIVATE_KEY_PASSWORD',
      'JACS_PASSWORD_FILE',
    ]);
    expect(fixture.private_key_password_strategy).toBe('single_source_required');

    const tmp = await mkdtemp(join(tmpdir(), 'haiai-node-init-contract-'));
    try {
      const keyDir = join(tmp, 'keys');
      const configPath = join(tmp, 'jacs.config.json');
      await mkdir(keyDir, { recursive: true });
      await writeFile(configPath, JSON.stringify({
        jacsAgentName: 'agent-alpha',
        jacsAgentVersion: '1.0.0',
        jacsKeyDir: './keys',
      }));
      const config = await loadConfig(configPath);

      const fileNames = fixture.private_key_candidate_order.map((name) =>
        name.replace('{agentName}', 'agent-alpha'),
      );
      const candidates = fileNames.map((name) => join(keyDir, name));
      await writeFile(candidates[0], 'first');
      await writeFile(candidates[1], 'second');
      await writeFile(candidates[2], 'third');

      await expect(loadPrivateKey(config)).resolves.toBe('first');

      await unlink(candidates[0]);
      await expect(loadPrivateKey(config)).resolves.toBe('second');

      await unlink(candidates[1]);
      await expect(loadPrivateKey(config)).resolves.toBe('third');
    } finally {
      await rm(tmp, { recursive: true, force: true });
    }
  });

  it('bootstrap register contract matches shared fixture', async () => {
    const fixture = loadInitContractFixture();
    const keypair = generateKeypair();
    const client = await HaiClient.fromCredentials(
      'bootstrap-agent',
      keypair.privateKeyPem,
      { url: 'https://hai.example', privateKeyPassphrase: 'keygen-password' },
    );

    const registerMock = vi.fn(async (options: Record<string, unknown>) => {
      expect(options.owner_email).toBe('owner@hai.ai');
      expect(options.domain).toBe('agent.example');
      return {
        agent_id: 'agent-123',
        jacs_id: 'bootstrap-agent',
        registration_id: 'reg-1',
        registered_at: '2026-01-01T00:00:00Z',
      };
    });
    client._setFFIAdapter(createMockFFI({ registerNewAgent: registerMock }));

    await client.registerNewAgent('bootstrap-agent', {
      ownerEmail: 'owner@hai.ai',
      domain: 'agent.example',
      description: 'Node shared init contract',
      password: 'keygen-password',
      quiet: true,
    });
  });
});
